File: solutions/julia/role-playing-game/1/role-playing-game.jl
StringOrMissing = Union{String, Missing} # type union definition
IntOrNothing = Union{Int64, Nothing} # type union definition

@kwdef mutable struct Player
    name::StringOrMissing = missing
    level::Int64 = 0
    health::Int64 = 100
    mana::IntOrNothing = nothing
end

function introduce(player::Player)
    if ismissing(player.name)
        return "Mighty Magician"
    end
    player.name
end

function increment(mana::IntOrNothing)
    if isnothing(mana)
        return 50
    else
        return mana + 100
    end
end

function increment(name::StringOrMissing)
    if ismissing(name)
        return "The Great"
    else
        return name * " the Great"
    end
end

function title!(player::Player)
    if player.level == 42
        player.name = increment(player.name)
    end
    player.name
end

function revive!(player::Player)
    if player.health == 0
        player.health = 100
        player.mana = increment(player.mana)
    end
    player
end

File: solutions/fsharp/role-playing-game/1/RolePlayingGame.fs
module RolePlayingGame

type Player =
    { Name: string option
      Level: int
      Health: int
      Mana: int option }

let introduce (player: Player) : string =
    match player.Name with
    | Some n -> n
    | None -> "Mighty Magician"

let revive (player: Player) : Player option =
    match player.Health with
    | 0 when player.Level >= 10 ->
        Some
            { player with
                Health = 100
                Mana = Some 100 }
    | 0 -> Some { player with Health = 100 }
    | _ -> None

let castSpell (manaCost: int) (player: Player) : Player * int =
    match player.Mana with
    | Some m when m >= manaCost ->
        ({ player with
            Mana = Some(m - manaCost) },
         2 * manaCost)
    | None ->
        ({ player with
            Health = max 0 (player.Health - manaCost) }),
        0
    | _ -> (player, 0)
